File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/verticalTraversal.ts
import {arrayToBTLevelOrder, BTNode} from "./BinaryTreeOperations";

interface Tuple {
	node: BTNode;
	vertical: number;
	level: number;
}

function verticalTraversal(root: BTNode) {
	let q: Tuple[] = [];
	let nodes: { [key: number]: { [key: number]: number[] } } = {};
	q.push({node: root, vertical: 0, level: 0});
	while (q.length > 0) {
		let {node, vertical, level} = q.shift();
		if(!nodes[vertical]) {
			nodes[vertical] = {};
		}
		if (!nodes[vertical][level]) {
			nodes[vertical][level] = [];
		}
		nodes[vertical][level].push(node.data);
		if (node.left) {
			q.push({node: node.left, vertical: vertical - 1, level: level + 1});
		}
		if (node.right) {
			q.push({node: node.right, vertical: vertical + 1, level: level + 1});
		}
	}
	let ans = [];
	const sortedVerticals = Object.keys(nodes).sort((a, b) => Number(a) - Number(b));

	for (const vertical of sortedVerticals) {
		const sortedLevels = Object.keys(nodes[vertical]).sort((a, b) => Number(a) - Number(b));
		const col: number[] = [];

		for (const level of sortedLevels) {
			col.push(...nodes[vertical][level]);
		}

		ans.push(col);
	}
	console.log(ans)
}

let root = arrayToBTLevelOrder([1, 2, 3, 4, 5, 6, 7]);
verticalTraversal(root);

File: StriverA2Z/Java/BinaryTrees/VerticalTraversal.java
import com.sun.source.tree.Tree;

import java.util.*;

public class VerticalTraversal {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[]{1, 2, 3, 4, 5, 6, 7});
        List<List<Integer>> ans = verticalTraversal(root);
        for (List<Integer> a : ans) {
            for (Integer i : a) {
                System.out.print(i + " ");
            }
        }
    }

    public static List<List<Integer>> verticalTraversal(Node root) {
        Queue<Tuple> q = new LinkedList<>();
        TreeMap<Integer, TreeMap<Integer, PriorityQueue<Integer>>> nodes = new TreeMap<>();
        q.offer(new Tuple(root, 0, 0));
        while (!q.isEmpty())  {
            Tuple tuple = q.poll();
            Node node = tuple.node;
            int vertical = tuple.vertical;
            int level = tuple.level;
            if(!nodes.containsKey(vertical)) {
                nodes.put(vertical, new TreeMap<>());
            }
            if (!nodes.get(vertical).containsKey(level)) {
                nodes.get(vertical).put(level, new PriorityQueue<>());
            }
            nodes.get(vertical).get(level).offer(node.data);
            if (node.left != null) {
                q.offer(new Tuple(node.left, vertical - 1, level + 1));
            }
            if (node.right != null) {
                q.offer(new Tuple(node.right, vertical + 1, level + 1));
            }
        }
        List<List<Integer>> list = new ArrayList<>();
        for (TreeMap<Integer, PriorityQueue<Integer>> mpp : nodes.values()) {
            list.add(new ArrayList<>());
            for (PriorityQueue<Integer> j : mpp.values()) {
                while (!j.isEmpty()) {
//                    System.out.print(j.peek());
                    list.get(list.size() - 1).add(j.poll());
                }
            }
        }
        return list;
    }
}
